redshift_sqlalchemy/dialect.py

```diff
@@ -320,67 +323,157 @@ def visit_unload_from_select(element, compiler, **kw):
-           """ % \
-           {'schema_name': element.schema_name,
-            'table_name': element.table_name,
-            'data_location': element.data_location,
-            'access_key': element.access_key,
-            'secret_key': element.secret_key,
-            'session_token': ';token=%s' % element.session_token if element.session_token else '',
-            'null': ("NULL '%s'" % element.options.get('null')) if element.options.get('null') else '',
-            'delimiter': element.options.get('delimiter', ','),
-            'ignore_header': element.options.get('ignore_header', 0),
-            'manifest': 'MANIFEST' if bool(element.options.get('manifest', False)) else '',
-            'empty_as_null': 'EMPTYASNULL' if bool(element.options.get('empty_as_null', True)) else '',
-            'blanks_as_null': 'BLANKSASNULL' if bool(element.options.get('blanks_as_null', True)) else ''}
+    qs = """COPY :schema.:table FROM :data_location
+    CREDENTIALS :credentials
+    :format
+    TRUNCATECOLUMNS
+    DELIMITER :delimiter
+    IGNOREHEADER :ignore_header
+    NULL :null
+    :manifest
+    :compression
+    :empty_as_null
+    :blanks_as_null
+    """
+
+    return compiler.process(
+        sa.text(qs).bindparams(
+            sa.bindparam('schema',
+                         value=element.table.schema or 'public',
+                         type_=_UnquotedParameter),
+            sa.bindparam('table',
+                         value=element.table.name,
+                         type_=_UnquotedParameter),
+            sa.bindparam('data_location',
+                         value=element.data_location,
+                         type_=sa.String),
+            sa.bindparam('credentials', value=element.credentials,
+                         type_=sa.String),
+            sa.bindparam('null', value=element.null, type_=sa.String),
+            sa.bindparam('delimiter',
+                         value=element.delimiter,
+                         type_=sa.String()),
+            sa.bindparam('ignore_header',
+                         value=element.ignore_header,
+                         type_=sa.Integer),
+            sa.bindparam('manifest',
+                         value='MANIFEST' if element.manifest else '',
+                         type_=_UnquotedParameter),
+            sa.bindparam('empty_as_null',
+                         value='EMPTYASNULL' if element.empty_as_null else '',
+                         type_=_UnquotedParameter),
+            sa.bindparam('blanks_as_null',
+                         value='BLANKSASNULL' if element.blanks_as_null else '',
+                         type_=_UnquotedParameter),
+            sa.bindparam('format',
+                         value=element.format,
+                         type_=_UnquotedParameter),
+            sa.bindparam('compression',
+                         value=element.compression,
+                         type_=_UnquotedParameter)
+        ),
+        literal_binds=True
+    )
```

Comment:
I mean to say using .format has the same security considerations as
bindparams when you disable quoting.
On 24 Aug 2015 23:24, "Phillip Cloud" <notifications@github.com> wrote:

> In redshift_sqlalchemy/dialect.py
> <https://github.com/graingert/redshift_sqlalchemy/pull/21#discussion_r37812363>
> :
>
> > +                         type_=_UnquotedParameter),
> > +            sa.bindparam('empty_as_null',
> > +                         value='EMPTYASNULL' if element.empty_as_null else '',
> > +                         type_=_UnquotedParameter),
> > +            sa.bindparam('blanks_as_null',
> > +                         value='BLANKSASNULL' if element.blanks_as_null else '',
> > +                         type_=_UnquotedParameter),
> > +            sa.bindparam('format',
> > +                         value=element.format,
> > +                         type_=_UnquotedParameter),
> > +            sa.bindparam('compression',
> > +                         value=element.compression,
> > +                         type_=_UnquotedParameter)
> > +        ),
> > +        literal_binds=True
> > +    )
>
> How so? All of the bind params are either checked for proper values or
> correctly quoted. Can you give an example with this PR of an injection
> that's still possible?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/graingert/redshift_sqlalchemy/pull/21/files#r37812363>
> .
>